File: clients/subc-client-swift/Sources/SubcClient/ConnectionFile.swift
import Foundation

// Port of subc-transport's connection_file.rs reader (and the TS mirror in
// connection-file.ts). The connection file is the daemon's published rendezvous
// record; its `key` is the shared transport secret. We refuse to trust a key
// from a file other local users can read (owner-only 0600), so a leaked key is a
// loud failure rather than a silent downgrade.

public let SCHEMA_VERSION = 1
public let MIN_KEY_LEN = 32
public let DAEMON_ID_LEN = 16

public struct Endpoint {
    public let host: String
    public let port: UInt16
}

public struct ConnectionInfo {
    public let schema: Int
    public let endpoints: [Endpoint]
    public let key: Data
    public let daemonId: Data
    public let pid: Int
    public let daemonVer: String
}

public struct ConnectionFileError: Error { public let message: String }

public func readConnectionFile(_ path: String) throws -> ConnectionInfo {
    try verifyOwnerOnly(path)

    let raw = try Data(contentsOf: URL(fileURLWithPath: path))
    guard let obj = try JSONSerialization.jsonObject(with: raw) as? [String: Any] else {
        throw ConnectionFileError(message: "connection file is not a JSON object: \(path)")
    }

    guard let schema = obj["schema"] as? Int else {
        throw ConnectionFileError(message: "connection file missing integer 'schema'")
    }
    guard schema == SCHEMA_VERSION else {
        throw ConnectionFileError(message: "unsupported connection file schema \(schema); expected \(SCHEMA_VERSION)")
    }

    guard let endpointsRaw = obj["endpoints"] as? [[String: Any]] else {
        throw ConnectionFileError(message: "connection file 'endpoints' must be an array")
    }
    let endpoints = try endpointsRaw.map { e -> Endpoint in
        guard let host = e["host"] as? String, let port = e["port"] as? Int else {
            throw ConnectionFileError(message: "endpoint must be { host: string, port: number }")
        }
        return Endpoint(host: host, port: UInt16(port))
    }
    guard !endpoints.isEmpty else {
        throw ConnectionFileError(message: "connection file must include at least one endpoint")
    }

    let key = try bytes(obj["key"], "key")
    let daemonId = try bytes(obj["daemon_id"], "daemon_id")
    guard key.count >= MIN_KEY_LEN else {
        throw ConnectionFileError(message: "connection file key too short: \(key.count) bytes, need >= \(MIN_KEY_LEN)")
    }
    guard daemonId.count == DAEMON_ID_LEN else {
        throw ConnectionFileError(message: "daemon_id must be \(DAEMON_ID_LEN) bytes, got \(daemonId.count)")
    }

    return ConnectionInfo(
        schema: schema,
        endpoints: endpoints,
        key: key,
        daemonId: daemonId,
        pid: obj["pid"] as? Int ?? 0,
        daemonVer: obj["daemon_ver"] as? String ?? ""
    )
}

private func bytes(_ value: Any?, _ field: String) throws -> Data {
    guard let arr = value as? [Int] else {
        throw ConnectionFileError(message: "connection file field '\(field)' must be a JSON array of bytes")
    }
    return Data(arr.map { UInt8(truncatingIfNeeded: $0) })
}

/// On unix, reject any group/other permission bit: the key is published
/// owner-only (0600), so a wider mode means the secret has leaked.
private func verifyOwnerOnly(_ path: String) throws {
    let attrs = try FileManager.default.attributesOfItem(atPath: path)
    guard let perm = (attrs[.posixPermissions] as? NSNumber)?.intValue else { return }
    if (perm & 0o077) != 0 {
        throw ConnectionFileError(message: "connection file \(path) has insecure permissions 0o\(String(perm, radix: 8)); expected owner-only 0600")
    }
}

File: clients/subc-client/src/connection-file.ts
// Port of subc-transport's connection_file.rs reader. The connection file is
// the daemon's published rendezvous record; its `key` is the shared transport
// secret. We refuse to trust a key from a file other local users can read,
// exactly as the Rust reader does, so a leaked (group/world-readable) key is a
// loud failure rather than a silent downgrade.

import { promises as fs } from "node:fs";

import { PROTOCOL_VERSION } from "./envelope.js";

export const SCHEMA_VERSION = 1;
export const MIN_KEY_LEN = 32;
export const DAEMON_ID_LEN = 16;

export interface Endpoint {
  host: string;
  port: number;
}

export interface ConnectionInfo {
  schema: number;
  endpoints: Endpoint[];
  /** Transport key bytes. Serialized on disk as a JSON array of numbers. */
  key: Uint8Array;
  /** 16-byte daemon identity. Serialized on disk as a JSON array of numbers. */
  daemonId: Uint8Array;
  pid: number;
  daemonVer: string;
}

export class ConnectionFileError extends Error {}

function toBytes(value: unknown, field: string): Uint8Array {
  if (!Array.isArray(value)) {
    throw new ConnectionFileError(`connection file field '${field}' must be a JSON array of bytes`);
  }
  for (const byte of value) {
    if (typeof byte !== "number" || !Number.isInteger(byte) || byte < 0 || byte > 255) {
      throw new ConnectionFileError(`connection file field '${field}' has invalid byte ${String(byte)}`);
    }
  }
  return Uint8Array.from(value as number[]);
}

function validate(info: ConnectionInfo): void {
  if (info.schema !== SCHEMA_VERSION) {
    throw new ConnectionFileError(
      `unsupported connection file schema ${info.schema}; expected ${SCHEMA_VERSION}`,
    );
  }
  if (info.endpoints.length === 0) {
    throw new ConnectionFileError("connection file must include at least one endpoint");
  }
  if (info.key.length < MIN_KEY_LEN) {
    throw new ConnectionFileError(
      `connection file key is too short: ${info.key.length} bytes, need at least ${MIN_KEY_LEN}`,
    );
  }
  if (info.daemonId.length !== DAEMON_ID_LEN) {
    throw new ConnectionFileError(
      `connection file daemon_id must be ${DAEMON_ID_LEN} bytes, got ${info.daemonId.length}`,
    );
  }
}

/**
 * On unix, reject any group/other permission bit: the key is published
 * owner-only (0600), so a wider mode means the secret has leaked. On Windows the
 * file inherits the per-user profile directory ACL at create time and there are
 * no portable mode bits to re-check, matching the Rust no-op.
 */
async function verifyOwnerOnly(path: string): Promise<void> {
  if (process.platform === "win32") return;
  const stat = await fs.stat(path);
  const mode = stat.mode & 0o777;
  if ((mode & 0o077) !== 0) {
    throw new ConnectionFileError(
      `connection file ${path} has insecure permissions 0o${mode.toString(8)}; expected owner-only 0600`,
    );
  }
}

/** Read, permission-check, and validate a connection file. */
export async function readConnectionFile(path: string): Promise<ConnectionInfo> {
  await verifyOwnerOnly(path);
  const raw = await fs.readFile(path, "utf8");
  let parsed: Record<string, unknown>;
  try {
    parsed = JSON.parse(raw) as Record<string, unknown>;
  } catch (err) {
    throw new ConnectionFileError(`connection file JSON read failed for ${path}: ${String(err)}`);
  }

  const wireVersion = parsed.wire_version;
  if (wireVersion !== undefined && wireVersion !== PROTOCOL_VERSION) {
    throw new ConnectionFileError(
      `connection file wire_version ${String(wireVersion)} but this client speaks ${PROTOCOL_VERSION}; the client library must be upgraded`,
    );
  }

  const endpointsRaw = parsed.endpoints;
  if (!Array.isArray(endpointsRaw)) {
    throw new ConnectionFileError("connection file 'endpoints' must be an array");
  }
  const endpoints: Endpoint[] = endpointsRaw.map((e) => {
    const ep = e as Record<string, unknown>;
    if (typeof ep.host !== "string" || typeof ep.port !== "number") {
      throw new ConnectionFileError("connection file endpoint must be { host: string, port: number }");
    }
    return { host: ep.host, port: ep.port };
  });

  const info: ConnectionInfo = {
    schema: parsed.schema as number,
    endpoints,
    key: toBytes(parsed.key, "key"),
    daemonId: toBytes(parsed.daemon_id, "daemon_id"),
    pid: parsed.pid as number,
    daemonVer: (parsed.daemon_ver as string) ?? "",
  };
  validate(info);
  return info;
}
